Package good practice guide (#3829)
umbraco/UmbracoDocs: Extending/Packages/Good-practice/index.md

Some folders within Umbraco will already exist for all installations, if you access these folders you need to be aware of the case used to ensure you end up in the right place: 

| Folder | Note | 
|-|-|
| /App_Plugins | Uppercase `A` and `P` |
| /App_Pluigns/[Ll]ang | Uppercase `L` but in later version of Umbraco 9.3+ can be upper or lowercase `L` |
| /Views | Uppercase `V` | 
| /umbraco/Licenses | Lowercase `u` uppercase `L` |
| /config | Lowercase `c` |

:::note 
If you create a custom section/tree umbraco will also build paths based on the name of that section or tree, these folder paths will also be case sensitive: 

For example If you have a custom tree with the treeAlias of `MyCustomTree` umbraco will look for files in `App_Plugins\MyCustomTree\backoffice\MyCustomTree\` 
:::

### Path manipulation
Building folder path strings manually can cause problems when swapping between file systems. Windows uses the backslash character to '\' separate folders and files, Linux uses the forward slash '/'.

You should use the .Net `Path` commands wherever possible when building paths to ensure that the correct path is build. 

``` cs
// WRONG: don't build paths manually.

> var myPath = myFolder + "\myfile.txt";

Author: I think i need to add something about getting paths via the FileProvider or `MapPath` commands, because it should never be hardwired, but this example could then have a mapped path value for 'myFolder' and it would make more sense.

---
versionFrom: 9.0.0
meta.Title: "Package development good practice"

> meta.Description: "Information on common defaults and good practice for umbraco package development"

Reviewer: ```suggestion
meta.Description: "Information on good practice and common defaults for Umbraco package development"
```

---

# Good practice and defaults
_This document provides some guides and notes on package development, and good practice that will help you maintain and support your package through multiple releases and versions of Umbraco. These good practices are not prescriptive, but mearly offer a guide as to what we have found works well, and not-so-well, when developing packages for Umbraco_

> ## App_Plugins

Reviewer: Should this also explain what does go into the App_Plugins folder? Something like: 'Each package should create a subfolder with the package name/id (e.g. `App_Plugins\MyPackage`) to store their backoffice related code, like a `package.manifest`, the AngularJS views/controllers, CSS and language files.'

And maybe also note that these files will be served publicly (even though this folder isn't located in the `wwwroot`)?

Files in the `App_Plugins` folder should be considered immutable, that is they are not something a user of your package should be expecting to change on their site. 

The default delivery method for files to the App_Plugins folder is via a `.targets` file within a package, this means when a website is built the files in this folder are copied over from the nuget cache, and any changes a user might have made will be lost. Equally if the user performs a `dotnet clean` on a solution all files in the App_Plugins folder will likely be deleted. 

If you have files that you expect users of your package to alter you should not place them in the `App_Plugins` folder.

## Views
Views are used to render content on the front end of a website. If your plugin provides a way for the user to present the content publicly you should copy these files to the views folder. 

As the files will still be copied during build you should ensure your target file doesn't overwrite newer or altered files and doesn't delete files on clean. 

> _This method isn't perfect and if you do have a package that is heavy on views you might consider building a dotnet template or Razor class library to contain the views so they don't pollute your user's solutions._ 

Reviewer: Another way is to use ASP.NET [Tag Helpers](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/intro) or [View Components](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/view-components) instead. Or maybe even better: include the views in a Razor Class Library instead?

Author: yes this is a better way (RCL) but until v10 this isn't smooth enough - i suspect events will overtake this document 😄 and we should change it to how to do this in a RCL.

Reviewer: RCL projects have only been properly supported since v11, so it makes sense to only include additional information about how and when to use that project type in a version 11 specific documentation page...

## License files
Umbraco products store their licenses in `/umbraco/Licences` and it is considered a good location for third party packages that require license files to also store their licence files there. 

_The default .gitignore for Umbraco templates will include any files in this folder in a repo (while ignoring most of the rest of the Umbraco folder)_

:::note
The `/umbraco/Licenses` folder does not exist on a fresh installation of Umbraco so you might need to create it before you save your license file to this folder. 
:::

## Operating system considerations
Umbraco 9 is a .NET Core application and it can be ran on multiple operating systems (e.g Windows, Linux). When developing packages there are a few things you should be aware of so your package can run on all possible operating systems. 

> ### Case sensitivity 

Reviewer: Is using constants to keep the casing the same something we should advise?

Author: Yes - added a line

The Linux file system is case sensitive, so `App_Plugins/myPackage` is a different location to `app_plugins/MyPACKAGE`. When building your package you should ensure that you always refer to folders and paths in a consistent way. 

:::tip
You can adjust the case sensitivity of a Windows folder by running a command against the folder. 

``` bash
fsutil.exe file queryCaseSensitiveInfo <path>
```
:::

### Well known folders and case sensitivity  
Some folders within Umbraco will already exist for all installations, if you access these folders you need to be aware of the case used to ensure you end up in the right place: 

| Folder | Note | 
|-|-|
| /App_Plugins | Uppercase `A` and `P` |
| /App_Plugins/[Ll]ang | Uppercase `L` but in later version of Umbraco 9.3+ can be upper or lowercase `L` |
| /Views | Uppercase `V` | 
| /umbraco/Licenses | Lowercase `u` uppercase `L` |

> | /config | Lowercase `c` |

Reviewer: This isn't used in v9 anymore (and the `umbraco\config\lang` folder should only include project specific/user language overrides, so shouldn't be something for package authors to deal with).

Author: i think the caseness of these folders is worth noting, and if you want to make a package that works on all v9s you need to know that pre v9.3 the case thing was an issue. (again if we make this a v10 doc, we can simplify)

:::note 
If you create a custom section/tree Umbraco will also build paths based on the name of that section or tree, these folder paths will also be case sensitive: 

For example if you have a custom tree with the treeAlias of `MyCustomTree` Umbraco will look for files in `App_Plugins\MyCustomTree\backoffice\MyCustomTree\` 
:::

### File/folder locations
You should never hardwire a file or folder location into code if possible. Instead you should either access files file the `IHostingEnvrionment` FileProviders or by using the built in methods to access well known locations 

#### Temp folder
The location of the Umbraco temp folder can be controlled via configuration and cannot be assumed, you should use the `IHostingEnvironment.LocalTempPath` variable to locate the temp folder. 

``` cs
var localTempRoot = Path.GetFullPath(hostingEnvironment.LocalTempPath);
```

#### Relative paths
If you require the path of a folder relative to the site root, you can use the `IHostingEnvironment` methods to map a path. 

``` cs 

> var contentRootPath = hostingEnvironment.MapPathContentRoot("/");
> var webrootPath = hostEnvironment.MapPathWebRoot("/");

Reviewer: This should also show how you can create an absolute path to use as URL (and maybe also explain virtual paths/application relative paths):
```suggestion
var contentRootPath = hostingEnvironment.MapPathContentRoot("~/"); // Full physical path to the content/project root
var webrootPath = hostEnvironment.MapPathWebRoot("~/"); // Full physical path to the web root (served publicly)
var absolutePath = hostEnvironment.ToAbsolute("~/"); // Absolute path to use as URL
```

```

### Folder Access 

> You should not assume things about the folder structure of a site, and ideally you should not use direct IO commands to access the file system. Within a Asp.Net Core site access to the disk is usually managed with FileProviders, you can get access to the file providers from the `IWebHostEnvironment` class.

Reviewer: ```suggestion
You should not assume things about the folder structure of a site, and ideally you should not use direct IO commands to access the file system. Within an ASP.NET Core site access to the disk is usually managed with FileProviders, you can get access to the file providers from the `IWebHostEnvironment` class.
```

```cs
public class MyController: UmbracoAuthorizedJsonController
{
    public MyController(IWebHostEnvironment webHostEnvironment)
    {
        var webRootProvider = webHostEnvironment.WebRootFileProvider;

        var myFileInfo = webRootProvider.GetFileInfo("robots.txt");
        if (myFileInfo.Exists)
        {
            using (var stream = myFileInfo.CreateReadStream())
            {
                using(var streamReader = new StreamReader(stream))
                {
                    var text = streamReader.ReadToEnd();
                }
            }
        }
    }
}
```

:::note
This is the preferred method for file IO. because not all files served up by a site may in fact be in the wwwroot folder when you expect them to be - this is especially true if the site is using Razor class library to insert static files.
:::

### Path manipulation
Building folder path strings manually can cause problems when swapping between file systems. Windows uses the backslash character to '\\' separate folders and files, Linux uses the forward slash '/'.

> You should use the .Net `Path` commands wherever possible when building paths to ensure that the correct path is built. 

Reviewer: ```suggestion
You should use the .NET `Path` commands wherever possible when building paths to ensure that the correct path is built. 
```

``` cs
// WRONG: don't build paths manually.
var myPath = webrootPath + "\robots.txt";

// CORRECT: build the path using the correct separator for the current file system.
var myPath = Path.Combine(webrootPath, "robots.txt"); 
```

If you do need to build a path manually use `Path.DirectorySeparatorChar` to get the correct separator for the filesystem.

## Settings
All but the most simple package will probably require some settings to be stored for the users to control and change the behavior of how the package works on a site. Where you store these settings will depend a lot on the nature of the package. 

### Property Editors
Property Editors should store their settings as part of their datatype in Umbraco. This is the standard way property editors behavior is controlled, it is familar to users and is supported by deployment tools. 

### Don't save to appsettings.json
You should not alter `appsettings.json` in code. 

> Settings in aspnet core are merged from a number of different locations at runtime. You cannot guarantee that `appsettings.json` is the location that a setting is read from and your users may well not want certain settings in that file. You can read settings from the configuration but you cannot assume they have come from `appsettings.json`.

Reviewer: ```suggestion
Settings in ASP.NET Core are merged from a number of different locations at runtime. You cannot guarantee that `appsettings.json` is the location that a setting is read from and your users may well not want certain settings in that file. You can read settings from the configuration, but you cannot assume they have come from `appsettings.json`.
```

### Settings locations
There are many options for where you might save your settings and a lot will depend on the nature of your package: 

> #### Save to database 

Reviewer 2: Maybe a section about the IKeyValueService in Umbraco to store simple settings. I use this in my RedirectsViewer package : https://github.com/dawoe/umbraco-redirectsviewer/blob/develop/Source/Our.Umbraco.RedirectsViewer/Controllers/UserGroupsApiController.cs#L89

Author: yeah - originally i did write a load of potential implementation options into the document, it got to big 😞 - so i went for streamlined, this document is more 'pointers' than actual implementation, not sure the balance is right, but maybe there should be other pages for this type of thing ?

Reviewer: I agree that we should this document from becoming too big, but maybe we can add at least a single mention of this, e.g. 'Simple values can also be stored in the database using the `IKeyValueService`' and link that to the API docs?

Author: I've added a line about IKeyValueService , but i am not sure there is actually a documentation page to link to ?
